rust/yarp/src/lib.rs

```diff
@@ -0,0 +1,254 @@
+//! # yarp
+//!
+//! Rustified version of Ruby's YARP parser.
+//!
+#![warn(
+    box_pointers,
+    clippy::all,
+    clippy::nursery,
+    clippy::pedantic,
+    future_incompatible,
+    missing_docs,
+    nonstandard_style,
+    rust_2018_idioms,
+    trivial_casts,
+    trivial_numeric_casts,
+    unreachable_pub,
+    unused_qualifications
+)]
+
+// Most of the code in this file is generated, so sometimes it generates code
+// that doesn't follow the clippy rules. We don't want to see those warnings.
+#![allow(
+    clippy::too_many_lines,
+    clippy::use_self
+)]
+
+mod bindings {
+    // In `build.rs`, we generate bindings based on the config.yml file. Here is
+    // where we pull in those bindings and make them part of our library.
+    include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
+}
+
+pub use self::bindings::*;
+
+use std::ffi::{CStr, CString};
+use std::mem::MaybeUninit;
+use yarp_sys::{yp_parser_t, yp_parser_init, yp_parse, yp_node_destroy, yp_node_t, yp_comment_t, yp_diagnostic_t};
+
+use std::ptr::NonNull;
+
+/// A diagnostic message that came back from the parser.
+#[derive(Copy, Clone, Debug)]
+pub struct Diagnostic(*mut yp_diagnostic_t);
```

Comment:
As as a tiny point of optimization, using the `NonNull` here would make sense here as well, since it would allow the [null-pointer optimization](https://www.sobyte.net/post/2022-04/rust-enum-layout/).